lib/uklock/include/uk/mutex.h

```diff
@@ -98,53 +100,71 @@ void uk_mutex_get_metrics(struct uk_mutex_metrics *dst);
 static inline void uk_mutex_lock(struct uk_mutex *m)
 {
 	struct uk_thread *current;
-	unsigned long irqf;
 
 	UK_ASSERT(m);
 
 	current = uk_thread_current();
 
+	/* if the owner is current thread then it definitely can't unlock during
+	 * uk_mutex_lock, hence if the owner is current thread then we can simply
+	 * increment the lock count
+	 */
+#ifdef CONFIG_LIBUKLOCK_MUTEX_RECURSE
+	if (unlikely(m->owner == current)) {
+		ukarch_inc(&(m->lock_count));
+		return;
+	}
+#else
+	UK_ASSERT(m->owner != current);
+#endif
+
+
 	for (;;) {
-		uk_waitq_wait_event(&m->wait,
-			m->lock_count == 0 || m->owner == current);
-		irqf = ukplat_lcpu_save_irqf();
-		if (m->lock_count == 0 || m->owner == current)
+		/* TODO: make wait queue SMP safe */
```

Comment:
Looking at the changes to the waitqueue below, I guess this is done by now?

Reply:
yes I added the a lock for a queue, but in future we need a SMP safe threads and scheduler